daemon/containerd/image_prune.go

```diff
@@ -106,18 +111,18 @@ func (i *ImageService) pruneUnused(ctx context.Context, filterFunc imageFilterFu
 		}
 	}
 
-	logrus.WithField("images", imagesToPrune).Debug("pruning")
-
+	possiblyDeletedConfigs := map[digest.Digest]struct{}{}
 	for _, img := range imagesToPrune {
-		blobs := []ocispec.Descriptor{}
+		logrus.WithField("image", img).Debug("pruning image")
 
-		err = containerdimages.Walk(ctx, presentChildrenHandler(store, containerdimages.HandlerFunc(
-			func(_ context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
-				blobs = append(blobs, desc)
-				return nil, nil
-			})),
-			img.Target)
+		blobs := []ocispec.Descriptor{}
 
+		err := i.walkPresentChildren(ctx, img.Target, func(desc ocispec.Descriptor) {
+			blobs = append(blobs, desc)
+			if containerdimages.IsConfigType(desc.MediaType) {
+				possiblyDeletedConfigs[desc.Digest] = struct{}{}
+			}
+		})
 		if err != nil {
 			errs = multierror.Append(errs, err)
 			continue
```

Comment:
It's not the biggest deal to loop through the whole collection unconditionally, though it'd accumulate a lot of `context.Err()` errors if the context becomes done partway through

Reply:
Added early returns for context cancellation errors.